### ISSUE TITLE
Look into supporting OpenSSL 3

### ISSUE THREAD

**Issue by jonasbn** [dsully/perl-crypt-openssl-pkcs12#40]
See: https://github.com/openssl/openssl/blob/master/doc/man7/migration_guide.pod

**Comment by xnox** [comment]
If one export openssl.cnf that loads legacy provider the module builds and works fine with openssl v3.

But if one disables legacy (default confiuration) things don't work.

Test keys, certs, and p12 files use obsolete encryptions which are not the default anymore. Even if I rebuild the test.p12 file with modern encryption / mac / salting algos as are default in openssl v3, the test suite still fails with many errors w.r.t. to invalid options and asn1 reoutines.

Please regenerate test data from scratch using large 4k+ RSA keys, certs with sha256+ checksums, sha256 mac, etc.

**Comment by real-dam** [comment]
Hi,
Any news? Debian is adopting OpenSSL 3 and this module becomes unreleasable.
I know the documentation says that only openssl 1.x is supported, but are there plans to remove that restriction?

Damyan,
Debian Perl Group

**Comment by real-dam** [comment]
Hi,
Re-creating the test certificates/p12 indeed helps. Here's a script that can be used for this:
```
#!/bin/sh

openssl req -nodes -x509 -days 10000 \
    -subj "/C=US/ST=California/O=ElectricRain/CN=test.electricrain.com" \
    -out test-cert.pem -keyout test-key.pem

openssl pkcs12 -export -in test-cert.pem -inkey test-key.pem -out test.p12 \
    -password pass:testing
```
It doesn't use the greatest key sizes etc, but the current openssl defaults. Still, `openssl pkcs12 -info -in test.p12` works and displays the certificate and the private key, so I assume the defaults are good enough. Perhaps not very future-proof, but OK.

With test data re-created this way, the test suite passes several tests and fails at changing the password:
```
t/pkcs12.t ......... 
1..16
ok 1 - use Crypt::OpenSSL::PKCS12;
# Attempting to read certificate from certs/test.p12
ok 2 - PKCS object created
ok 3 - PEM certificate created
ok 4 - Asserting PEM key
ok 5 - Asserting mac
ok 6 - Asserting PKCS12 as string
PKCS12_newpass: testing foo
C0D20349F87F0000:error:068000A8:asn1 encoding routines:asn1_check_tlen:wrong tag:../crypto/asn1/tasn_dec.c:1188:
C0D20349F87F0000:error:0688010A:asn1 encoding routines:asn1_d2i_ex_primitive:nested asn1 error:../crypto/asn1/tasn_dec.c:752:
C0D20349F87F0000:error:0688010A:asn1 encoding routines:asn1_template_noexp_d2i:nested asn1 error:../crypto/asn1/tasn_dec.c:685:Field=salt, Type=PBEPARAM
C0D20349F87F0000:error:0680006E:asn1 encoding routines:ASN1_item_unpack:decode error:../crypto/asn1/asn_pack.c:60:
C0D20349F87F0000:error:11800072:PKCS12 routines:PKCS12_newpass:parse error:../crypto/pkcs12/p12_npas.c:49:
not ok 7 - Changing password
```
I don't understand why, though. The PKCS12 struct is handled via openssl routines and it loads a file with current defaults which should be fine.

Skipping the tests before the password change (i.e., just load test.p12 and try to change the password right away) also fails.

I hope this gives some clues.

**Comment by timlegge** [comment]
The PR #42 sorts this.  Apparently openssl 3.x release versions have a bug that cannot change the password on a pkcs12 file.  Thinking of it I should ass a not to the pod for that PR

**Comment by timlegge** [comment]
Fixed in #42
